src/Crypto/WebAuthn/Operation/Authentication.hs

```diff
@@ -160,26 +157,28 @@ newtype AuthenticationResult = AuthenticationResult
 -- The 'arSignatureCounterResult' field of the result should be inspected to
 -- enforce Relying Party policy regarding potentially cloned authenticators.
 verifyAuthenticationResponse ::
+  (Monad m) =>
   -- | The origin of the server
   M.Origin ->
   -- | The hash of the relying party id
   M.RpIdHash ->
-  -- | The user handle, in case the user is identified already
-  -- TODO: Mention that this would be empty for username-less authentication
+  -- | The currently identified user. Set if the user was identified  before
+  -- the ceremony was initiated, e.g., via a username or cookie.
+  -- This should be set to 'Nothing' if you are implementing a login flow where
+  -- passkeys are automatically filled in by the browser in the login form.
   Maybe M.UserHandle ->
-  -- | The database entry for the credential, as created in the initial
-  -- attestation and optionally updated in subsequent assertions
-  CredentialEntry ->
+  -- | Function to look up the credential in the database. This function is used
+  -- to verify that the credential belongs to the user identified by either the
+  -- 'userIdentified' parameter if it is not 'Nothinbg' or the 'M.araUserHandle'
+  -- field of the response.
+  (M.UserHandle -> M.CredentialId -> m (Maybe CredentialEntry)) ->
   -- | The options that were passed to the get() method
   M.CredentialOptions 'M.Authentication ->
   -- | The credential returned from get()
   M.Credential 'M.Authentication 'True ->
-  -- | Either a non-empty list of validation errors in case of the assertion
-  -- being invalid
-  -- Or in case of success a signature counter result, which should be dealt
-  -- with
-  Validation (NonEmpty AuthenticationError) AuthenticationResult
-verifyAuthenticationResponse origin rpIdHash midentifiedUser entry options credential = do
+  -- | Either an error or the result of the authentication ceremony
+  m (Either AuthenticationError AuthenticationResult)
```

Comment:
Because I now have to do a monadic decision; I can't use `Validation` anymore.  This means we always short circuit on the first error instead of collecting errors.